packages/common/src/ui/layout/tables/context/TableContext.ts

```diff
@@ -6,6 +6,8 @@ export interface RowState {
   isSelected: boolean;
   isExpanded: boolean;
   isDisabled: boolean;
+  isFocused: boolean;
+  index: number;
```

Comment:
Yeah, that makes sense. So currently, it just lists them in the order provided by the back-end, right?

Just thinking this through, brain dump to follow...  -- when you make a change, new data is refetched after the mutation, which causes the table to get content in new order and re-render. That's when `setRows` runs, which re-renders a new list of rows.

Are you suggesting that it would still work this way, but we'd sort the new rows according their ID order in the sorted array rather than the order that they're supplied? And then only update the ID array (based on the current `data` order) when clicking an appropriate header? So if the "setRows" runs as a result of the `useEffect` on `data`  change, how do we know whether to update the the ID array or not (cos it's not clear whether the data changes because a new order was requested (by clicking column header) or because a value changed). I guess you could keep a "headerClicked" state value, but you might have a better suggestion.

Anyway, assuming that all works fine, just wondering what happens to rows that are either missing (from ID array) or for IDs in array that are no longer in the resulting rows (due to the value change causing a row to change pages). I guess you'd just put any new items at the bottom, and omit any that were missing, then update the ID array accordingly.

Reply:
Yeah that's all sounding right. 

I did a quick minimal sandbox of how the client side sorting in the detail views is currently working: https://codesandbox.io/s/vibrant-mountain-8qspzm?file=/src/App.tsx

It's not as bad in practise when you're using a modal to edit 😁 

> I guess you could keep a "headerClicked" state value,

This would essentially be the object that stores the current sort state, which when a header is clicked, would change.

>  guess you'd just put any new items at the bottom, and omit any that were missing, then update the ID array accordingly.

Yup .. annoying!

---------

Thanks for thinking so deeply about this, it made me think about it more also. I realised that we previously were not sending updates to the server so frequently. Now that we are, there actually is really no reason to use client side sorting..! In a later iteration, I think asking the backend to support sorting on the `lines` field on requisition/stocktake/invoices would be a good idea!